File: OldProblem/vnoi/ROTATEXT.py
import sys


data = sys.stdin.buffer.read().split()
ptr = 0


def read_int():
    global ptr
    value = int(data[ptr])
    ptr += 1
    return value


def read_str():
    global ptr
    value = data[ptr].decode()
    ptr += 1
    return value


n = read_int()
c = read_int()
q = read_int()
S = read_str()
T = read_str()
BASE = 256
MOD1, MOD2 = (1000000093, 1000000223)

pow1 = [1] * (n + 1)
pow2 = [1] * (n + 1)
same1 = [0] * (n + 1)
same2 = [0] * (n + 1)
for i in range(1, n + 1):
    pow1[i] = pow1[i - 1] * BASE % MOD1
    pow2[i] = pow2[i - 1] * BASE % MOD2
    same1[i] = (same1[i - 1] * BASE + 1) % MOD1
    same2[i] = (same2[i - 1] * BASE + 1) % MOD2

factors = []
value = c
div = 2
while div * div <= value:
    if value % div == 0:
        factors.append(div)
        while value % div == 0:
            value //= div
    div += 1
if value > 1:
    factors.append(value)


def find_root(mod):
    if c == 1:
        return 1
    step = (mod - 1) // c
    cand = 2
    while True:
        root = pow(cand, step, mod)
        if root != 1 and all(pow(root, c // factor, mod) != 1 for factor in factors):
            return root
        cand += 1


root1 = find_root(MOD1)
root2 = find_root(MOD2)
rot1 = [1] * c
rot2 = [1] * c
for i in range(1, c):
    rot1[i] = rot1[i - 1] * root1 % MOD1
    rot2[i] = rot2[i - 1] * root2 % MOD2
decode = {(rot1[i], rot2[i]): i for i in range(c)}

initial_s = [0] * (n + 1)
diff = [0] * (n + 1)
for i in range(1, n + 1):
    s_value = ord(S[i - 1]) - 97
    initial_s[i] = s_value
    diff[i] = (s_value - (ord(T[i - 1]) - 97)) % c

size = (n << 2) + 5
hd1 = [0] * size
hd2 = [0] * size
lazy = [0] * size
bit = [0] * (n + 2)


def bit_add(index, delta):
    while index <= n:
        bit[index] += delta
        index += index & -index


def add_s(left, right, delta):
    bit_add(left, delta)
    bit_add(right + 1, -delta)


def get_s(index):
    orig = index
    total = 0
    while index > 0:
        total += bit[index]
        index -= index & -index
    return (initial_s[orig] + total) % c


def apply(node, shift):
    if shift == 0:
        return
    hd1[node] = hd1[node] * rot1[shift] % MOD1
    hd2[node] = hd2[node] * rot2[shift] % MOD2
    lazy[node] += shift
    if lazy[node] >= c:
        lazy[node] -= c


def push(node):
    shift = lazy[node]
    if shift == 0:
        return
    left = node << 1
    apply(left, shift)
    apply(left | 1, shift)
    lazy[node] = 0


def pull(node, right_len):
    left = node << 1
    right = left | 1
    hd1[node] = (hd1[left] * pow1[right_len] + hd1[right]) % MOD1
    hd2[node] = (hd2[left] * pow2[right_len] + hd2[right]) % MOD2


def build(node, start, end):
    if start == end:
        value = diff[start]
        hd1[node] = rot1[value]
        hd2[node] = rot2[value]
        return

    mid = (start + end) >> 1
    left = node << 1
    build(left, start, mid)
    build(left | 1, mid + 1, end)
    pull(node, end - mid)


def update(node, start, end, left, right, shift):
    if start > right or end < left:
        return
    if left <= start and end <= right:
        apply(node, shift)
        return

    push(node)
    mid = (start + end) >> 1
    left_node = node << 1
    update(left_node, start, mid, left, right, shift)
    update(left_node | 1, mid + 1, end, left, right, shift)
    pull(node, end - mid)


def compare(node, start, end, left_bound):
    if end < left_bound:
        return 0
    seg_len = end - start + 1
    if left_bound <= start and hd1[node] == same1[seg_len] and hd2[node] == same2[seg_len]:
        return 0
    if start == end:
        delta = decode[(hd1[node], hd2[node])]
        s_char = get_s(start)
        t_char = (s_char - delta) % c
        if s_char < t_char:
            return -1
        return 1

    push(node)
    mid = (start + end) >> 1
    res = compare(node << 1, start, mid, left_bound)
    if res:
        return res
    return compare(node << 1 | 1, mid + 1, end, left_bound)


build(1, 1, n)

out = []
for _ in range(q):
    query_type = read_int()
    if query_type == 3:
        left = read_int()
        out.append("=><"[compare(1, 1, n, left)])
        continue

    left = read_int()
    right = read_int()
    shift = read_int() % c
    if shift == 0:
        continue
    if query_type == 1:
        add_s(left, right, shift)
        update(1, 1, n, left, right, shift)
    else:
        update(1, 1, n, left, right, (-shift) % c)

sys.stdout.write("\n".join(out))

File: OldProblem/vnoi/ROTATEXT.cpp
#include <bits/stdc++.h>

using namespace std;

using int64 = long long;

struct FastScanner {
    static constexpr size_t BUFSIZE = 1 << 20;
    char buf[BUFSIZE];
    size_t idx = 0;
    size_t size = 0;

    inline char read() {
        if (idx >= size) {
            size = fread(buf, 1, BUFSIZE, stdin);
            idx = 0;
            if (size == 0) {
                return 0;
            }
        }
        return buf[idx++];
    }

    template <class T>
    bool nextInt(T &out) {
        char c;
        do {
            c = read();
            if (!c) {
                return false;
            }
        } while (c <= ' ');

        T sign = 1;
        if (c == '-') {
            sign = -1;
            c = read();
        }

        T value = 0;
        while (c > ' ') {
            value = value * 10 + (c - '0');
            c = read();
        }
        out = value * sign;
        return true;
    }

    bool nextString(string &out) {
        char c;
        do {
            c = read();
            if (!c) {
                return false;
            }
        } while (c <= ' ');

        out.clear();
        while (c > ' ') {
            out.push_back(c);
            c = read();
        }
        return true;
    }
};

static inline int64 mulMod(int64 a, int64 b, int64 mod) {
    return static_cast<int64>((__int128)a * b % mod);
}

static int64 powMod(int64 base, int64 exp, int64 mod) {
    int64 result = 1;
    while (exp > 0) {
        if (exp & 1LL) {
            result = mulMod(result, base, mod);
        }
        base = mulMod(base, base, mod);
        exp >>= 1LL;
    }
    return result;
}

struct Fenwick {
    int n = 0;
    int mod = 1;
    vector<int> bit;

    Fenwick() = default;
    Fenwick(int n_, int mod_) : n(n_), mod(mod_), bit(n_ + 2, 0) {}

    void addPoint(int idx, int value) {
        if (value == 0) {
            return;
        }
        for (; idx <= n; idx += idx & -idx) {
            bit[idx] += value;
            if (bit[idx] >= mod) {
                bit[idx] -= mod;
            }
        }
    }

    void addRange(int left, int right, int value) {
        if (value == 0) {
            return;
        }
        addPoint(left, value);
        if (right + 1 <= n) {
            addPoint(right + 1, mod - value);
        }
    }

    int prefix(int idx) const {
        int result = 0;
        for (; idx > 0; idx -= idx & -idx) {
            result += bit[idx];
            if (result >= mod) {
                result -= mod;
            }
        }
        return result;
    }
};

struct Solver {
    static constexpr int64 BASE = 911382323LL;
    static constexpr pair<int64, int64> MODS[27] = {
        {0, 0},
        {1000000007LL, 1000000009LL},
        {1000000007LL, 1000000009LL},
        {1000000009LL, 1000000021LL},
        {1000000009LL, 1000000021LL},
        {1000000021LL, 1000000181LL},
        {1000000009LL, 1000000021LL},
        {1000000009LL, 1000000093LL},
        {1000000009LL, 1000000033LL},
        {1000000009LL, 1000000207LL},
        {1000000021LL, 1000000181LL},
        {1000000123LL, 1000000321LL},
        {1000000009LL, 1000000021LL},
        {1000000093LL, 1000000223LL},
        {1000000009LL, 1000000093LL},
        {1000000021LL, 1000000321LL},
        {1000000033LL, 1000000097LL},
        {1000000181LL, 1000000453LL},
        {1000000009LL, 1000000207LL},
        {1000000021LL, 1000000097LL},
        {1000000021LL, 1000000181LL},
        {1000000009LL, 1000000093LL},
        {1000000123LL, 1000000321LL},
        {1000000349LL, 1000000579LL},
        {1000000009LL, 1000000033LL},
        {1000000801LL, 1000000901LL},
        {1000000093LL, 1000000223LL},
    };

    int n = 0;
    int c = 0;
    int q = 0;
    int64 mod1 = 0;
    int64 mod2 = 0;
    string s;
    string t;
    vector<int64> pow1, pow2, same1, same2, rot1, rot2;
    vector<int> initialS;
    vector<int64> h1, h2;
    vector<int> lazy;
    Fenwick bit;

    int decodeDelta(int64 x1, int64 x2) const {
        for (int d = 0; d < c; ++d) {
            if (rot1[d] == x1 && rot2[d] == x2) {
                return d;
            }
        }
        return 0;
    }

    int64 findRoot(int64 mod) const {
        if (c == 1) {
            return 1;
        }

        vector<int> factors;
        int value = c;
        for (int div = 2; div * div <= value; ++div) {
            if (value % div == 0) {
                factors.push_back(div);
                while (value % div == 0) {
                    value /= div;
                }
            }
        }
        if (value > 1) {
            factors.push_back(value);
        }

        const int64 step = (mod - 1) / c;
        for (int64 cand = 2;; ++cand) {
            int64 root = powMod(cand, step, mod);
            if (root == 1) {
                continue;
            }
            bool ok = true;
            for (int factor : factors) {
                if (powMod(root, c / factor, mod) == 1) {
                    ok = false;
                    break;
                }
            }
            if (ok) {
                return root;
            }
        }
    }

    void apply(int node, int shift) {
        if (shift == 0) {
            return;
        }
        h1[node] = mulMod(h1[node], rot1[shift], mod1);
        h2[node] = mulMod(h2[node], rot2[shift], mod2);
        lazy[node] += shift;
        if (lazy[node] >= c) {
            lazy[node] -= c;
        }
    }

    void push(int node) {
        int shift = lazy[node];
        if (shift == 0) {
            return;
        }
        int left = node << 1;
        apply(left, shift);
        apply(left | 1, shift);
        lazy[node] = 0;
    }

    void pull(int node, int rightLen) {
        int left = node << 1;
        int right = left | 1;
        h1[node] = (mulMod(h1[left], pow1[rightLen], mod1) + h1[right]) % mod1;
        h2[node] = (mulMod(h2[left], pow2[rightLen], mod2) + h2[right]) % mod2;
    }

    void build(int node, int left, int right, const vector<int> &diff) {
        if (left == right) {
            h1[node] = rot1[diff[left]];
            h2[node] = rot2[diff[left]];
            return;
        }

        int mid = (left + right) >> 1;
        int leftNode = node << 1;
        build(leftNode, left, mid, diff);
        build(leftNode | 1, mid + 1, right, diff);
        pull(node, right - mid);
    }

    void update(int node, int start, int end, int left, int right, int shift) {
        if (start > right || end < left) {
            return;
        }
        if (left <= start && end <= right) {
            apply(node, shift);
            return;
        }

        push(node);
        int mid = (start + end) >> 1;
        int leftNode = node << 1;
        update(leftNode, start, mid, left, right, shift);
        update(leftNode | 1, mid + 1, end, left, right, shift);
        pull(node, end - mid);
    }

    int getCurrentS(int index) const {
        int value = initialS[index] + bit.prefix(index);
        if (value >= c) {
            value -= c;
        }
        return value;
    }

    int compare(int node, int start, int end, int leftBound) {
        if (end < leftBound) {
            return 0;
        }

        int len = end - start + 1;
        if (leftBound <= start && h1[node] == same1[len] && h2[node] == same2[len]) {
            return 0;
        }

        if (start == end) {
            int delta = decodeDelta(h1[node], h2[node]);
            int sChar = getCurrentS(start);
            int tChar = sChar - delta;
            if (tChar < 0) {
                tChar += c;
            }
            return (sChar < tChar) ? -1 : 1;
        }

        push(node);
        int mid = (start + end) >> 1;
        int res = compare(node << 1, start, mid, leftBound);
        if (res != 0) {
            return res;
        }
        return compare(node << 1 | 1, mid + 1, end, leftBound);
    }

    void run() {
        FastScanner fs;
        fs.nextInt(n);
        fs.nextInt(c);
        fs.nextInt(q);
        fs.nextString(s);
        fs.nextString(t);

        tie(mod1, mod2) = MODS[c];
        pow1.assign(n + 1, 1);
        pow2.assign(n + 1, 1);
        same1.assign(n + 1, 0);
        same2.assign(n + 1, 0);
        for (int i = 1; i <= n; ++i) {
            pow1[i] = mulMod(pow1[i - 1], BASE, mod1);
            pow2[i] = mulMod(pow2[i - 1], BASE, mod2);
            same1[i] = (mulMod(same1[i - 1], BASE, mod1) + 1) % mod1;
            same2[i] = (mulMod(same2[i - 1], BASE, mod2) + 1) % mod2;
        }

        rot1.assign(c, 1);
        rot2.assign(c, 1);
        int64 root1 = findRoot(mod1);
        int64 root2 = findRoot(mod2);
        for (int i = 1; i < c; ++i) {
            rot1[i] = mulMod(rot1[i - 1], root1, mod1);
            rot2[i] = mulMod(rot2[i - 1], root2, mod2);
        }

        initialS.assign(n + 1, 0);
        vector<int> diff(n + 1, 0);
        for (int i = 1; i <= n; ++i) {
            int sValue = s[i - 1] - 'a';
            int tValue = t[i - 1] - 'a';
            initialS[i] = sValue;
            diff[i] = sValue - tValue;
            if (diff[i] < 0) {
                diff[i] += c;
            }
        }

        h1.assign((n << 2) + 5, 0);
        h2.assign((n << 2) + 5, 0);
        lazy.assign((n << 2) + 5, 0);
        bit = Fenwick(n, c);
        build(1, 1, n, diff);

        string out;
        out.reserve(static_cast<size_t>(q) * 2);
        for (int i = 0; i < q; ++i) {
            int type;
            fs.nextInt(type);
            if (type == 3) {
                int left;
                fs.nextInt(left);
                int cmp = compare(1, 1, n, left);
                out.push_back(cmp < 0 ? '<' : (cmp > 0 ? '>' : '='));
                out.push_back('\n');
                continue;
            }

            int left, right, shift;
            fs.nextInt(left);
            fs.nextInt(right);
            fs.nextInt(shift);
            shift %= c;
            if (shift == 0) {
                continue;
            }

            if (type == 1) {
                bit.addRange(left, right, shift);
                update(1, 1, n, left, right, shift);
            } else {
                update(1, 1, n, left, right, (c - shift) % c);
            }
        }

        if (!out.empty()) {
            out.pop_back();
        }
        fwrite(out.data(), 1, out.size(), stdout);
    }
};

constexpr pair<int64, int64> Solver::MODS[27];

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    Solver solver;
    solver.run();
    return 0;
}
